src/transformers/models/mlcd/modular_mlcd.py

```diff
@@ -0,0 +1,599 @@
+
+        # Select and flatten the embeddings based on the position IDs
+        rotary_pos_emb = rotary_pos_emb_full[pos_ids].flatten(1)
+
+        return rotary_pos_emb
+
+
+class MLCDVisionEmbeddings(CLIPVisionEmbeddings):
+    def __init__(self, config: MLCDVisionConfig):
+        super().__init__(config)
+        del self.position_embedding
+
+    def forward(self, pixel_values: torch.FloatTensor) -> torch.Tensor:
+        batch_size = pixel_values.shape[0]
+        target_dtype = self.patch_embedding.weight.dtype
+        # patch_embeds -> shape = [batch, width, grid, grid]
+        patch_embeds = self.patch_embedding(pixel_values.to(dtype=target_dtype))
+        patch_embeds = patch_embeds.flatten(2).transpose(1, 2)
+
+        class_embeds = self.class_embedding.expand(batch_size, 1, -1)
+        embeddings = torch.cat([class_embeds, patch_embeds], dim=1)
+
+        return embeddings
+
+
+class MLCDAttention(CLIPAttention):
+    """Multi-headed attention with RoPE. Refer to papers:
+    - Attention is all you need:
+        https://arxiv.org/abs/1706.03762
+    - RoFormer: Enhanced Transformer with Rotary Position Embedding:
+        https://arxiv.org/abs/2104.09864
+    """
+
+    def __init__(self, config: MLCDVisionConfig):
+        super().__init__(config)
+        self.num_key_value_groups = config.num_key_value_groups
+        self.is_causal = False
+
+    def forward(
+        self,
+        hidden_states: torch.Tensor,
+        rotary_pos_emb: Optional[torch.Tensor] = None,
+        position_embeddings: Optional[Tuple[torch.Tensor, torch.Tensor]] = None,
+        attention_mask: Optional[torch.Tensor] = None,
+        **kwargs: Unpack[FlashAttentionKwargs],
+    ) -> Tuple[torch.Tensor, Optional[torch.Tensor]]:
+        batch_size, seq_length = hidden_states.shape[:-1]
+
+        # Each of shape: [batch_size, seq_length, num_heads, head_dim]
+        query_states = self.q_proj(hidden_states).reshape((batch_size, seq_length, self.num_heads, self.head_dim))
+        key_states = self.k_proj(hidden_states).reshape((batch_size, seq_length, self.num_heads, self.head_dim))
+        value_states = self.v_proj(hidden_states).reshape((batch_size, seq_length, self.num_heads, self.head_dim))
+        if position_embeddings is None:
+            logger.warning_once(
+                "The attention layers in this model are transitioning from computing the RoPE embeddings internally "
+                "through `rotary_pos_emb` (2D tensor of RoPE theta values), to using externally computed "
+                "`position_embeddings` (Tuple of tensors, containing cos and sin). In v4.54 `rotary_pos_emb` will be "
+                "removed and `position_embeddings` will be mandatory."
+            )
+            emb = torch.cat((rotary_pos_emb, rotary_pos_emb), dim=-1)
+            cos = emb.cos().unsqueeze(0).float()
+            sin = emb.sin().unsqueeze(0).float()
+        else:
+            cos, sin = position_embeddings
```

Comment:
Which way is actually used? Not sure if we need a warning for a new model, can we avoid it?

Reply:
@tanhuajie I mean, can we avoid `if position_embeddings is None` path completely?